rustls-mbedcrypto-provider/src/kx.rs

```diff
@@ -140,6 +169,101 @@ impl crypto::ActiveKeyExchange for KeyExchange {
     }
 }
 
+struct DheKxGroup {
+    name: NamedGroup,
+    group: ffdhe_groups::FfdheGroup<'static>,
+    priv_key_len: usize,
+}
+
+impl fmt::Debug for DheKxGroup {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{:?}", self.name)
+    }
+}
+
+impl SupportedKxGroup for DheKxGroup {
+    fn start(&self) -> Result<Box<dyn ActiveKeyExchange>, Error> {
+        let g = mbedtls::bignum::Mpi::from_binary(self.group.g).map_err(mbedtls_err_to_rustls_error)?;
+        let p = mbedtls::bignum::Mpi::from_binary(self.group.p).map_err(mbedtls_err_to_rustls_error)?;
+
+        let mut rng = super::rng::rng_new().ok_or(rustls::crypto::GetRandomFailed)?;
+        let mut x = vec![0; self.priv_key_len];
+        rng.random(&mut x)
+            .map_err(|_| rustls::crypto::GetRandomFailed)?;
+        let x = mbedtls::bignum::Mpi::from_binary(&x)
+            .map_err(|e| Error::General(format!("failed to make Bignum from random bytes: {}", e)))?;
+
+        let x_pub = g
+            .mod_exp(&x, &p)
+            .map_err(mbedtls_err_to_rustls_error)?;
+
+        Ok(Box::new(DheActiveKeyExchange {
+            name: self.name,
+            group: self.group,
+            x: x.to_binary()
+                .map_err(mbedtls_err_to_rustls_error)?,
+            x_pub: x_pub
+                .to_binary()
+                .map_err(mbedtls_err_to_rustls_error)?,
+        }))
+    }
+
+    fn name(&self) -> NamedGroup {
+        self.name
+    }
+}
+
+struct DheActiveKeyExchange {
+    name: NamedGroup,
+    group: ffdhe_groups::FfdheGroup<'static>,
+    x: Vec<u8>,
+    x_pub: Vec<u8>,
+}
+
+impl ActiveKeyExchange for DheActiveKeyExchange {
+    fn complete(self: Box<Self>, peer_pub_key: &[u8]) -> Result<crypto::SharedSecret, Error> {
+        let y_pub = Mpi::from_binary(peer_pub_key).map_err(mbedtls_err_to_rustls_error)?;
+
+        let x = Mpi::from_binary(&self.x).map_err(mbedtls_err_to_rustls_error)?;
+        let p = Mpi::from_binary(self.group.p).map_err(mbedtls_err_to_rustls_error)?;
+
+        let one = Mpi::new(1).map_err(mbedtls_err_to_rustls_error)?;
+
+        let mut p_minus_one = p;
+        p_minus_one -= &one;
+
+        // https://www.rfc-editor.org/rfc/rfc7919.html#section-5.1:
+        // Peers MUST validate each other's public key Y [...] by ensuring that 1 < Y < p-1.
+        if !(one < y_pub && y_pub < p_minus_one) {
+            return Err(Error::General(
+                "Invalid DHE key exchange public key received; pub key must be in range (1, p-1)".into(),
```

Comment:
Would be good to have a negative unit case test for this.

Reply:
Good suggestion!

Reply:
Done!